scheduler cache: API and implementation

hongchaodeng opened the pull request:
This is the cache interface, impl. and tests separated out from #20669.
It depends on #20977.

hongchaodeng committed:
scheduler cache: API and implementation

hongchaodeng commented:
/cc @wojtek-t

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,237 @@
+		return nil
+	}
+
+	key := mustGetPodKey(pod)
```
Moving discussion from parent PR here:

@wojtekt
"I'm against having "mustGetPodKey" function. We shouldn't have any "panic" in the code in places where it depends on the data (there can be some data-corruption it shouldn't blow up our components).
It should be:

key, err := getPodKey(pod)
if err != nil {
return err
}

And you can move it outside lock BTW."

@hongchaodeng 
"To return error is not intended, because this error case is not transient error (e.g. data corruption). The pod must have a unique key, e.g. a name.

Currently, Pod.Name is unique within Pod.Namespace, and we get "Namespace/Name" as key by using MetaNamespaceKeyFunc. If some changes breaks this compatibility, it is very dangerous and we shouldn't swallow the error.

We can extract namespace and name from api.Pod directly. But depending on MetaNamespaceKeyFunc seems more maintainable.

Not quite sure about this part. Just let you know that we shouldn't let such error not handled."

wojtek-t replied:
I don't agree with you here.

Basically, assume that you have a corrupted data in etcd. Then:
- if we leave as you are suggesting (with panic), this will cause scheduler crash-loop
- if we change to what I'm suggesting, we may end up with scheduler trying to schedule that pod multiple pods, but it will be scheduling other pods in the meantime too.
The second option is much better, since it doesn't break the system completely.

So I would like this to be changed to return an error.

hongchaodeng replied:
Just point out one thing because I didn't know why corrupted data can cause the panic.

api.Pod is ensured from compiler level to have [GetObjectMeta()](https://github.com/kubernetes/kubernetes/blob/9030f16071f4f6d46f0482995f0db5eaf96b9294/pkg/api/register.go#L116-L116) and be instance of [ObjectMetaAccessor](https://github.com/kubernetes/kubernetes/blob/c1d932e44a1a16890791df4f5d3f0d5f75b13b53/pkg/api/meta/interfaces.go#L31-L31). Thus, MetaNamespaceKeyFunc should always return the key. See [here](https://github.com/kubernetes/kubernetes/blob/8f203a28f1d17b9371673ff147aa7e8f34f093f5/pkg/client/cache/store.go#L79-L79) and [here](https://github.com/kubernetes/kubernetes/blob/8f203a28f1d17b9371673ff147aa7e8f34f093f5/pkg/api/meta/meta.go#L34-L34).

It's orthogonal to returning error. I agree with you that It's fine to return error. Just to sync on why I think the panic is for "compatibility".

wojtek-t replied:
Yes, I agree that we should always be able to access all necessary fields. But we can e.g. change "MetaNamespaceKeyFunc" to do some other kind of validation, and it can potentially return an error.

So I would suggest changing to return an error.

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,237 @@
+	// This mutex guards all fields within this cache struct.
+	mu          sync.Mutex
+	assumedPods map[string]assumedPod
+	podStates   map[string]podState
```
Moving from parent PR discussion:

@wojtekt:
"I think it would be cleaner to have
pods map[string]podInfo

podInfo:
pod *api.Pod
state podState
deadline time.Time

that to have two maps (that are slightly duplicating the information). WDYT?
 @wojtek-t  kubernetes member
wojtek-t added a note 22 hours ago
Or even we will not need state in such case, because we can define it as:
deadline = 0 <=> pod is added
deadline != 0 <=> pod is assumed
WDYT?"

@hongchaodeng :

""assumedPods" is expected to be a much smaller subset of pods. It's separated out to do efficient scanning and cleaning up."

wojtek-t replied:
To be honest, I don't think that lookup in small maps comparing to bigger maps makes such a difference.
But OK, I but this argument.

However, in that case I would suggest to changing to something like that:

1. podStates map[string]podState

podState struct {
  pod *api.Pod
  deadline time.Time
}
deadline = 0 <=> pod is added
deadline != <=> pod is assumed

2. assumedPods    sets.String

Then you have all information about pods in a single map (podStates), and you additionally have fast lookup whether the pod is assumed or not.

WDYT?

wojtek-t commented on plugin/pkg/scheduler/schedulercache/interface.go:
```diff
@@ -0,0 +1,81 @@
+
+// Cache collects pods' information and provides node-level aggregated information.
+// It's intended for generic scheduler to do efficient lookup.
+// Cache's operations are pod centric. It incrementally updates itself based on pod event.
```
s/event/events/

wojtek-t commented on plugin/pkg/scheduler/schedulercache/interface.go:
```diff
@@ -0,0 +1,81 @@
+// It's intended for generic scheduler to do efficient lookup.
+// Cache's operations are pod centric. It incrementally updates itself based on pod event.
+// Pod events are sent via network.
+// We don't have guaranteed delivery of all events. The current informer uses DeltaQueue and Reflector.
```
s/uses DeltaQueue and Reflector/uses Reflector/
[The Reflector is what matters here].

wojtek-t commented:
@hongchaodeng 
I've added some comments. Once you apply them and fix tests, I will take a look again (but generally it looks ok).

hongchaodeng commented:
Addressed all comments. The test failure was caused by the dependency on #20977. I have picked it here to show the test result.

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
```
nit: remove empty line

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+func (cache *schedulerCache) List(selector labels.Selector) ([]*api.Pod, error) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
```
nit: remove empty line

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+		// Pod was expired. We should add it back.
+		cache.addPod(pod)
+	default:
+		return fmt.Errorf("pod state wasn't assumed or expired but get added. Pod key: %v", key)
```
In fact this branch means that it was "ok & !cache.assumedPods", which mean that pod was already in added state.

So how about changing error message to:
"pod was already in added state. Pod key: %v", key
?

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+		err := cache.updatePod(oldPod, newPod)
+		if err != nil {
+			return err
+		}
```
nit:
if err := cache.updatePod(oldPod, newPod); err != nil {
  return err
}

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+	err := cache.deletePod(oldPod)
+	if err != nil {
+		return err
+	}
```
if err := cache.deletePod(oldPod); err != nil {
  return err
}

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+func (cache *schedulerCache) deletePod(pod *api.Pod) error {
+	n := cache.nodes[pod.Spec.NodeName]
+	err := n.removePod(pod)
+	if err != nil {
```
if err := n.removePod(pod); err ! =nil {

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+
+	_, ok := cache.podStates[key]
+	switch {
+	case ok && !cache.assumedPods[key]:
```
Is this correct? If the pod is assumed, its resources are also added to the cache. Shouldn't we remove them here?

wojtek-t replied:
ahh - looking into your state machine it seems that I'm wrong.
So the pattern you would like to have is that in case of assumed pod, we simply ignore the Remove event, and wait for its expiration?
Can you please add a comment about it (so that noone will think that it's a bug in the future)?

hongchaodeng replied:
An assumed pod won't have Delete/Remove event. It needs to have Add event before Remove event, in which case the state would change from Assumed to Added.

hongchaodeng replied:
Will add a comment.

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+
+func (cache *schedulerCache) expirePod(key string, ps *podState) error {
+	err := cache.deletePod(ps.pod)
+	if err != nil {
```
if err := cache.deletePod(ps.pod); err != nil {

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,268 @@
+		}
+		if now.After(ps.deadline) {
+			err := cache.expirePod(key, ps)
+			if err != nil {
```
if err := cache.expirePod(key, ps); err != nil {

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+
+	tests := []struct {
+		pods []*api.Pod
+
```
remove empty line

hongchaodeng replied:
It would be better if we separate out test input and expected result.
Let's leave it here

wojtek-t replied:
I still would prefer not having it, but well... If you really prefer, let's leave it

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+	testPods := []*api.Pod{
+		makeBasePod(nodeName, "test", "100m", "500", []api.ContainerPort{{HostPort: 80}}),
+		makeBasePod(nodeName, "test-1", "100m", "500", []api.ContainerPort{{HostPort: 80}}),
+		makeBasePod(nodeName, "test-2", "200m", "1Ki", []api.ContainerPort{{HostPort: 8080}}),
```
please make one of the pods to be 0-request pod (so that you also test nonzeroRequest)

hongchaodeng replied:
Added.

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+	for i, tt := range tests {
+		cache := newSchedulerCache(time.Second, time.Second, nil)
+		for _, pod := range tt.pods {
+			err := cache.AssumePodIfBindSucceed(pod, alwaysTrue)
```
if err := cache...(); err != nil {

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+	ttl := 10 * time.Second
+	tests := []struct {
+		pods        []*api.Pod
+		assumedTime []time.Time // assume time for individual pods
```
it's not the best idea to have two slices and require to both of them have the same size.
Can you please reuse podState for it? I mane:

tests := []struct {
  pods []podState
  cleanupTime time.Time
  wNodeInfo *NodeInfo
}

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+		pods        []*api.Pod
+		assumedTime []time.Time // assume time for individual pods
+		cleanupTime time.Time
+
```
remove empty line

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+		cache := newSchedulerCache(ttl, time.Second, nil)
+
+		for j, pod := range tt.pods {
+			err := cache.assumePodIfBindSucceed(pod, alwaysTrue, tt.assumedTime[j])
```
if err := ...; err != nil {

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+	basePod := makeBasePod(nodeName, "test", "100m", "500", []api.ContainerPort{{HostPort: 80}})
+	tests := []struct {
+		podToAssume *api.Pod
+		podToAdd    *api.Pod
```
Please generalize both to be []*api.Pod (to make more sophisticated test possible)

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+		wNodeInfo *NodeInfo
+	}{{ // Pod is assumed. It should be confirmed.
+		podToAssume: basePod,
+		podToAdd:    basePod,
```
Also, can you please change the test to have 2 pods to assume, and only one of them to be added?

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+	basePod := makeBasePod(nodeName, "test", "100m", "500", []api.ContainerPort{{HostPort: 80}})
+	tests := []struct {
+		podToAssume *api.Pod
+		podToAdd    *api.Pod
```
Please generalize both to be []*api.Pod (to make more sophisticated test possible)

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+	tests := []struct {
+		podToAssume *api.Pod
+		podToAdd    *api.Pod
+
```
remove empty line

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+	now := time.Now()
+	for i, tt := range tests {
+		cache := newSchedulerCache(ttl, time.Second, nil)
+		err := cache.assumePodIfBindSucceed(tt.podToAssume, alwaysTrue, now)
```
if err := ... ; err != nil {

here and in other places in this file

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+		podToAdd     *api.Pod
+		podToUpdate  *api.Pod
+		podToUpdate2 *api.Pod
+
```
remove empty line

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,467 @@
+		podToAssume  *api.Pod
+		podToAdd     *api.Pod
+		podToUpdate  *api.Pod
+		podToUpdate2 *api.Pod
```
can you please generalize it to be slices of pods instead?

wojtek-t commented:
@hongchaodeng - I added a bunch of different comments, but most of them are very minor now. Generally the logic looks ok now, so hopefully it should be ready soon. Thanks a lot!

hongchaodeng commented:
@wojtek-t Addressed your comments. Thanks for the suggestion.

hongchaodeng commented:
Rebased after #20977.

hongchaodeng commented:
@wojtek-t 
Can you rerun the test? I just added a comment and it seemed flaky failure.

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,264 @@
+	// This mutex guards all fields within this cache struct.
+	mu sync.Mutex
+	// a set of assumed pod keys.
+	// The key could further be  used to get an entry in podStates.
```
nit: S/be  /be / (remove one space - there are two spaces now)

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache.go:
```diff
@@ -0,0 +1,264 @@
+		}
+		if now.After(ps.deadline) {
+			if err := cache.expirePod(key, ps); err != nil {
+				glog.Errorf("expirePod failed: %v", err)
```
nit: expirePod failed for %s: %v", key, err

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,484 @@
+	}
+	tests := []struct {
+		// 0, 1: pod to assume
+		// 2: pod to add
```
I don't like this pattern. Can you please make:

podsToAssume []*api.Pod
podsToAdd        []*api.Pod
?
Or sth like that?

If at some point we decide to add another testcase with e.g. 3 pods, we will need to rewrite the whole test...

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,484 @@
+	nodeName := "node"
+	ttl := 10 * time.Second
+	basePod := makeBasePod(nodeName, "test", "100m", "500", []api.ContainerPort{{HostPort: 80}})
+	tests := []struct {
```
Please remove this struct here.Since you are testing a single specific scenario here, we don't want to have a loop for it here (as someone may want to add another testcase in the future, which basically doesn't make sense for this test).

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,484 @@
+	}
+	tests := []struct {
+		// 0: pod to assume and add
+		// 1, 2: pod to update
```
As above...
I don't like this pattern. Can you please make:

podsToAssume []api.Pod
podsToAdd []api.Pod
?
Or sth like that?

If at some point we decide to add another testcase with e.g. 3 pods, we will need to rewrite the whole test...

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,484 @@
+func TestRemovePod(t *testing.T) {
+	nodeName := "node"
+	basePod := makeBasePod(nodeName, "test", "100m", "500", []api.ContainerPort{{HostPort: 80}})
+	tests := []struct {
```
Similar to above - you are testing a specific single scenario here - it doesn't make sense to have an array of structs here (this is misleading)

wojtek-t commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -68,29 +72,84 @@ func (n *NodeInfo) RequestedResource() Resource {
+	n.nonzeroRequest.MilliCPU -= non0_cpu
+	n.nonzeroRequest.Memory -= non0_mem
+
+	k1, err := getPodKey(pod)
```
Please move it to the top - you don't want to change the underlying structure in that case, in case of error.

wojtek-t commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -68,29 +72,84 @@ func (n *NodeInfo) RequestedResource() Resource {
+	for i := range n.pods {
+		k2, err := getPodKey(n.pods[i])
+		if err != nil {
+			return err
```
change to log error an continue

wojtek-t commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -68,29 +72,84 @@ func (n *NodeInfo) RequestedResource() Resource {
+			// delete the element
+			n.pods[i] = n.pods[len(n.pods)-1]
+			n.pods = n.pods[:len(n.pods)-1]
+			break
```
return nil

wojtek-t commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -68,29 +72,84 @@ func (n *NodeInfo) RequestedResource() Resource {
+			break
+		}
+	}
+	return nil
```
return fmt.Errorf("no corresponding pod in pods")

[or some error like that]

wojtek-t commented:
@hongchaodeng - hopefull last pass of comments (just very minor for the code, and some other for test)

hongchaodeng commented:
@wojtek-t 
Addressed most of the comments. 

Regarding the table testing, I found it more clear to read and extend test like that. I like to keep it for consistency. But I am fine with either ways. If you insist, I can change it.

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,483 @@
+
+	for i, tt := range tests {
+		cache := newSchedulerCache(ttl, time.Second, nil)
+		if err := cache.assumePodIfBindSucceed(tt.podsToAssume[0], alwaysTrue, now); err != nil {
```
let's change it to loop over all podsToAssume

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,483 @@
+		if err := cache.assumePodIfBindSucceed(tt.podsToAssume[1], alwaysTrue, now); err != nil {
+			t.Fatalf("assumePod failed: %v", err)
+		}
+		if err := cache.AddPod(tt.podsToAdd[0]); err != nil {
```
let's change it to loop over all podsToAdd

wojtek-t commented on plugin/pkg/scheduler/schedulercache/cache_test.go:
```diff
@@ -0,0 +1,483 @@
+	now := time.Now()
+	for i, tt := range tests {
+		cache := newSchedulerCache(ttl, time.Second, nil)
+		if err := cache.assumePodIfBindSucceed(tt.podsToAssume[0], alwaysTrue, now); err != nil {
```
let's change it to loop over all: podsToAssume, podsToAdd, podsToUpdate

wojtek-t commented:
@hongchaodeng - thanks a lot for your patience. I've added few minor comments again, but I promise this is the last time. Once they are fixed, I will apply lgtm label.
Thanks!

hongchaodeng commented:
@wojtek-t 

Addressed comments
Thanks for your review too.

wojtek-t commented:
LGTM - thanks!

wojtek-t commented:
@k8s-bot e2e test this please github issue: #21463 #21467

wojtek-t commented:
@k8s-bot unit test this please github issue: #21451

wojtek-t commented:
@k8s-bot e2e test this please github issue: #20787

wojtek-t commented:
@k8s-bot e2e test this please github issue: #21467
